imap_processing/ultra/l2/ultra_l2.py

```diff
@@ -0,0 +1,334 @@
+"""Calculate ULTRA Level 2 (L2) ENA Map Product."""
+
+from __future__ import annotations
+
+import logging
+from pathlib import Path
+
+import numpy as np
+import xarray as xr
+
+from imap_processing.cdf.utils import load_cdf
+from imap_processing.ena_maps import ena_maps
+from imap_processing.ena_maps.utils.coordinates import CoordNames
+from imap_processing.ena_maps.utils.map_properties import (
+    DEFAULT_ULTRA_L2_MAP_PROPERTIES,
+    MapProperties,
+)
+from imap_processing.spice import geometry
+
+logger = logging.getLogger(__name__)
+logger.info("Importing ultra_l2 module")
+
+# Set some default values for the map properties
+PSET_SPICE_FRAME = geometry.SpiceFrame.IMAP_DPS
+DEFAULT_L2_HEALPIX_NSIDE = 32
+DEFAULT_L2_HEALPIX_NESTED = False
+DEFAULT_L2_MAP_PROPERTIES = DEFAULT_ULTRA_L2_MAP_PROPERTIES
+
+# These variables must always be present in each L1C dataset
+REQUIRED_L1C_VARIABLES = [
+    "counts",
+    "exposure_factor",
+    "sensitivity",
+    "background_rates",
+]
+
+# These variables are projected to the map as the mean of pointing set pixels value,
+# weighted by that pointing set pixel's exposure and solid angle
+VARIABLES_TO_WEIGHT_BY_POINTING_SET_EXPOSURE_TIMES_SOLID_ANGLE = [
+    "sensitivity",
+    "background_rates",
+    "observation_time",
+]
+
+# These variables are dropped after they are used to calculate flux and flux uncertainty
+# They will not be present in the final map
+VARIABLES_TO_DROP_AFTER_FLUX_CALCULATION = [
+    "counts",
+    "background_rates",
+    "pointing_set_exposure_times_solid_angle",
+    "num_pointing_set_pixel_members",
+    "corrected_count_rate",
+]
+
+
+def read_into_pointing_set(
+    input_data: xr.Dataset | str | Path, inplace: bool = False
+) -> ena_maps.UltraPointingSet:
+    """
+    Read a path or Dataset into an UltraPointingSet.
+
+    Parameters
+    ----------
+    input_data : xr.Dataset | str | Path
+        Path to the CDF file or xarray Dataset containing the L1C dataset.
+    inplace : bool
+        If True and if input_data is a Dataset, modify input_data in place.
+        Default is False, in which case a copy is made.
+        Does not affect str or Path type input_data.
+
+    Returns
+    -------
+    ena_maps.UltraPointingSet
+        An UltraPointingSet object containing the L1C dataset.
+
+    Raises
+    ------
+    ValueError
+        If input_data is neither an xarray Dataset nor a path to a CDF file.
+    KeyError
+        If any of the required variables are missing from the input data.
+    """
+    # Allow for passing in EITHER xarray Datasets (preferable for testing)
+    if isinstance(input_data, xr.Dataset):
+        if not inplace:
+            input_data = input_data.copy(deep=True)
+        ultra_pointing_set = ena_maps.UltraPointingSet(
+            l1c_dataset=input_data, spice_reference_frame=PSET_SPICE_FRAME
```

Comment:
Can `UltraPointingSet()` default to PSET_SPICE_FRAME so you don't have to pass it in explicitly in all of these paths here?